random/imx6ull-rngb.c

```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+
+static int systemInit(void)
+{
+	const char devpath[] = "random";
+	int err;
+	oid_t dev;
+
+	common_rngb.base = mmap(NULL, MMAP_SIZE, PROT_READ | PROT_WRITE, MAP_DEVICE | MAP_UNCACHED, OID_PHYSMEM, RNGB_START_ADDRESS);
+	if (common_rngb.base == MAP_FAILED) {
+		printf("rngb: could not map addr\n");
+		return -1;
+	}
+
+	err = portCreate(&common_rngb.port);
+	if (err != EOK) {
+		printf("rngb: could not create port\n");
+		systemCleanup(1);
+		return -1;
+	}
+
+	dev.port = common_rngb.port;
+	dev.id = 0;
+
+	err = mutexCreate(&common_rngb.lock);
+	if (err != EOK) {
+		printf("rngb: could not create mutex for rngb\n");
+		systemCleanup(2);
+		return -1;
+	}
+
+	err = condCreate(&common_rngb.cond);
+	if (err != EOK) {
+		printf("rngb: could not create cond for rngb\n");
+		systemCleanup(3);
+		return -1;
+	}
+
+	err = create_dev(&dev, devpath);
+	if (err != EOK) {
+		printf("rngb: could not create port file %s (err %d)\n", devpath, err);
+		systemCleanup(4);
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int hardwareInit(void)
+{
+	int err;
+
+	/* Reset and wait until rngb is sleeping */
+	*(common_rngb.base + rng_cmd) = 0x40;
+	do {
+		err = *(common_rngb.base + rng_sr) & 0x4;
+	} while (err == 0);
+
+	interrupt(RNGB_IRQ, rngb_intr, NULL, common_rngb.cond, NULL);
+
+	/* Run self test */
+	*(common_rngb.base + rng_cmd) = 0x1;
+	mutexLock(common_rngb.lock);
+	while (common_rngb.intr_st == 0) {
+		condWait(common_rngb.cond, common_rngb.lock, 3000);
+	}
+	mutexUnlock(common_rngb.lock);
+
+	if ((common_rngb.intr_st & ERROR_OCCURRED) != 0) {
+		printf("rngb: self-test failed\n");
+
+		systemCleanup(4);
+		return -1;
+	}
+
+	common_rngb.intr_st = 0;
+
+	/* Run seeding */
+	*(common_rngb.base + rng_cmd) = 0x2;
+	mutexLock(common_rngb.lock);
+	while (common_rngb.intr_st == 0) {
+		condWait(common_rngb.cond, common_rngb.lock, 3000);
+	}
+	mutexUnlock(common_rngb.lock);
+
+	if ((common_rngb.intr_st & ERROR_OCCURRED) != 0) {
+		printf("rngb: error after generating seed\n");
+
+		systemCleanup(4);
+		return -1;
+	}
+
+	common_rngb.intr_st = 0;
+
+	/* Enable automatic seeding */
+	*(common_rngb.base + rng_cr) = 0x10;
+	return 0;
+}
+
+
+static int readRandoms(char *buff, size_t size, unsigned mode)
+{
+	union {
+		uint32_t val32;
+		char val8[sizeof(uint32_t)];
+	} reg;
+	unsigned int i;
+	int err, level;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	mutexLock(common_rngb.lock);
+	for (i = 0; i < size; i++) {
+		err = *(common_rngb.base + rng_sr) & (1 << 16);
+		if (err != 0) {
+			mutexUnlock(common_rngb.lock);
+			return -EIO;
+		}
+
+		if (i % sizeof(uint32_t) == 0) {
+			level = (*(common_rngb.base + rng_sr) >> 8) & 0xf;
+			if (level == 0) {
+				if ((mode & O_NONBLOCK) != 0) {
+					break;
+				}
+
+				/* wait for new values in FIFO */
+				i--;
+				continue;
+			}
+
+			reg.val32 = *(common_rngb.base + rng_out);
+		}
+
+		buff[i] = reg.val8[i % sizeof(uint32_t)];
+	}
+	mutexUnlock(common_rngb.lock);
```

Comment:
I see two problems in the code:
1. You are checking the status register ( whether an error occurred ) for each byte, and you should in case if the output register will be read.
2. You a modyfing the iterator in the body `for` loop and this is bad practice and inconsistent with the MISRA rules (MISRA2012 14.2). It seems that a better solution would be to use a `while` loop here.
My proposal:
```suggestion
	unsigned int status;

	mutexLock(common_rngb.lock);
	i = 0;
	while (i < size) {
		if ((i % sizeof(uint32_t)) == 0) {
			status = *(common_rngb.base + rng_sr);
			if ((status & (1 << 16)) != 0) {
				mutexUnlock(common_rngb.lock);
				return -EIO;
			}
			if (((status >> 8) & 0xf) == 0) {
				if ((mode & O_NONBLOCK) != 0) {
					break;
				}

				/* wait for new values in FIFO */
				continue;
			}

			reg.val32 = *(common_rngb.base + rng_out);
		}

		buff[i] = reg.val8[i % sizeof(uint32_t)];
		i++;
	}
	mutexUnlock(common_rngb.lock);
```